File: scripts/artifact-tree-manifest.rb
#!/usr/bin/ruby

require 'digest'
require 'json'

abort('Usage: scripts/artifact-tree-manifest.rb PATH') unless ARGV.length == 1
root = File.expand_path(ARGV.fetch(0))
root_metadata = File.lstat(root)
abort('Artifact root must be an unsymlinked directory') unless root_metadata.directory? && !root_metadata.symlink?
entries = []

visit = lambda do |absolute, relative|
  metadata = File.lstat(absolute)
  display = relative.empty? ? '.' : relative
  mode = format('%04o', metadata.mode & 0o7777)
  if metadata.symlink?
    target = File.readlink(absolute)
    resolved_target = File.expand_path(target, File.dirname(absolute))
    unless resolved_target == root || resolved_target.start_with?("#{root}/")
      abort("Artifact symlink escapes root: #{display} -> #{target}")
    end
    entries << { path: display, type: 'symlink', mode: mode, target: target }
  elsif metadata.directory?
    entries << { path: display, type: 'directory', mode: mode }
    Dir.children(absolute).sort { |left, right| left.b <=> right.b }.each do |name|
      child_relative = relative.empty? ? name : "#{relative}/#{name}"
      visit.call(File.join(absolute, name), child_relative)
    end
  elsif metadata.file?
    entries << {
      path: display,
      type: 'file',
      mode: mode,
      size: metadata.size,
      sha256: Digest::SHA256.file(absolute).hexdigest
    }
  else
    abort("Unsupported artifact entry type: #{display}")
  end
end

visit.call(root, '')
STDOUT.write(JSON.generate({ version: 1, entries: entries }) + "\n")

File: scripts/artifact-tree-manifest.mjs
#!/usr/bin/env node

import { createHash } from 'node:crypto';
import { lstat, readFile, readdir, readlink } from 'node:fs/promises';
import path from 'node:path';
import process from 'node:process';

const rootArgument = process.argv[2];
if (!rootArgument || process.argv.length !== 3) {
  console.error('Usage: scripts/artifact-tree-manifest.mjs <path>');
  process.exit(2);
}

const root = path.resolve(rootArgument);

function bytewiseCompare(left, right) {
  return Buffer.compare(Buffer.from(left), Buffer.from(right));
}

function modeString(mode) {
  return (mode & 0o7777).toString(8).padStart(4, '0');
}

async function sha256(filePath) {
  return createHash('sha256').update(await readFile(filePath)).digest('hex');
}

const entries = [];

async function visit(absolutePath, relativePath) {
  const metadata = await lstat(absolutePath);
  const displayPath = relativePath || '.';
  if (metadata.isSymbolicLink()) {
    entries.push({ path: displayPath, type: 'symlink', mode: modeString(metadata.mode), target: await readlink(absolutePath) });
    return;
  }
  if (metadata.isDirectory()) {
    entries.push({ path: displayPath, type: 'directory', mode: modeString(metadata.mode) });
    const names = await readdir(absolutePath);
    names.sort(bytewiseCompare);
    for (const name of names) {
      await visit(path.join(absolutePath, name), relativePath ? `${relativePath}/${name}` : name);
    }
    return;
  }
  if (metadata.isFile()) {
    entries.push({
      path: displayPath,
      type: 'file',
      mode: modeString(metadata.mode),
      size: metadata.size,
      sha256: await sha256(absolutePath)
    });
    return;
  }
  throw new Error(`Unsupported artifact entry type: ${displayPath}`);
}

await visit(root, '');
process.stdout.write(`${JSON.stringify({ version: 1, entries })}\n`);
